File: src/main/java/org/omnifaces/component/script/ScriptErrorHandler.java
/*
 * Copyright OmniFaces
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package org.omnifaces.component.script;

import static java.lang.String.format;
import static org.omnifaces.config.OmniFaces.OMNIFACES_LIBRARY_NAME;
import static org.omnifaces.config.OmniFaces.OMNIFACES_SCRIPT_NAME;
import static org.omnifaces.util.FacesLocal.getRequestContextPath;
import static org.omnifaces.util.FacesLocal.getServletContext;
import static org.omnifaces.util.Utils.escapeJS;

import java.io.IOException;

import jakarta.faces.application.ResourceDependency;
import jakarta.faces.component.FacesComponent;
import jakarta.faces.context.FacesContext;
import jakarta.faces.event.ComponentSystemEvent;
import jakarta.faces.event.ListenerFor;
import jakarta.faces.event.PostAddToViewEvent;
import jakarta.servlet.ServletContext;

import org.omnifaces.cdi.ScriptError;
import org.omnifaces.config.OmniFaces;
import org.omnifaces.servlet.ScriptErrorServlet;
import org.omnifaces.util.Beans;
import org.omnifaces.util.State;

/**
 * <p>
 * The <code>&lt;o:scriptErrorHandler&gt;</code> is a component that catches uncaught JavaScript errors and unhandled
 * promise rejections on the client and sends them to the server via <code>navigator.sendBeacon()</code>, where they are
 * fired as {@link ScriptError} CDI events. This allows server-side logging of client-side JavaScript errors without
 * any additional endpoint boilerplate.
 *
 *
 * <h2 id="usage-client"><a href="#usage-client">Usage (client)</a></h2>
 * <p>
 * Just put the component in the <code>&lt;h:head&gt;</code> of the master template. It will automatically move itself to
 * <code>head</code> even when placed elsewhere, but placing it in <code>&lt;h:head&gt;</code> makes the intent explicit.
 * <pre>
 * &lt;h:head&gt;
 *     &lt;o:scriptErrorHandler /&gt;
 * &lt;/h:head&gt;
 * </pre>
 * <p>
 * This will register <code>window.onerror</code> and <code>unhandledrejection</code> event listeners which report
 * any uncaught client-side errors to the server. You can optionally specify a CSS selector via
 * <code>ignoreSelector</code> to suppress error reporting when a matching element exists in the document (e.g. on
 * error pages where errors are expected).
 * <pre>
 * &lt;o:scriptErrorHandler ignoreSelector="body.errorPage" /&gt;
 * </pre>
 *
 *
 * <h2 id="usage-server"><a href="#usage-server">Usage (server)</a></h2>
 * <p>
 * Create a CDI bean (typically application scoped) that observes the {@link ScriptError} event.
 * <pre>
 * &#64;ApplicationScoped
 * public class ScriptErrorObserver {
 *
 *     private static final Logger logger = Logger.getLogger(ScriptErrorObserver.class.getName());
 *
 *     public void onScriptError(&#64;Observes ScriptError error) {
 *         logger.warning(error.toString());
 *     }
 * }
 * </pre>
 * <p>
 * The {@link ScriptError} event provides the following information:
 * <ul>
 * <li>{@link ScriptError#pageURL()}: the URL of the page where the error occurred.</li>
 * <li>{@link ScriptError#errorMessage()}: the error message.</li>
 * <li>{@link ScriptError#sourceURL()}: the URL of the script file where the error occurred.</li>
 * <li>{@link ScriptError#lineNumber()}: the line number in the script.</li>
 * <li>{@link ScriptError#columnNumber()}: the column number in the script.</li>
 * <li>{@link ScriptError#errorName()}: the error type (e.g. "TypeError", "ReferenceError").</li>
 * <li>{@link ScriptError#errorStack()}: the full stack trace.</li>
 * <li>{@link ScriptError#remoteAddr()}: the remote address.</li>
 * <li>{@link ScriptError#userAgent()}: the User-Agent header.</li>
 * <li>{@link ScriptError#userPrincipal()}: the authenticated user principal name.</li>
 * </ul>
 *
 *
 * <h2 id="deduplication"><a href="#deduplication">Deduplication</a></h2>
 * <p>
 * To prevent flooding the server with repeated errors (e.g. from a <code>requestAnimationFrame</code> loop), the
 * client-side script deduplicates errors by message, source URL, and line number. The deduplication queue size and
 * expiry time can be configured via the <code>maxRecentErrors</code> and <code>errorExpiry</code> attributes.
 *
 * @author Bauke Scholtz
 * @see ScriptError
 * @see ScriptErrorServlet
 * @since 5.2
 */
@FacesComponent(value = ScriptErrorHandler.COMPONENT_TYPE, namespace = OmniFaces.OMNIFACES_NAMESPACE)
@ListenerFor(systemEventClass = PostAddToViewEvent.class)
@ResourceDependency(library = OMNIFACES_LIBRARY_NAME, name = OMNIFACES_SCRIPT_NAME, target = "head")
public class ScriptErrorHandler extends ScriptFamily {

    // Public constants -----------------------------------------------------------------------------------------------

    /** The component type, which is {@value org.omnifaces.component.script.ScriptErrorHandler#COMPONENT_TYPE}. */
    public static final String COMPONENT_TYPE = "org.omnifaces.component.script.ScriptErrorHandler";

    // Private constants ----------------------------------------------------------------------------------------------

    private static final int DEFAULT_MAX_RECENT_ERRORS = 100;
    private static final int DEFAULT_ERROR_EXPIRY = 60000;

    private static final String SCRIPT_INIT = "OmniFaces.ScriptErrorHandler.init('%s',%s,%s,%s);";

    private static final String ERROR_SERVLET_NOT_REGISTERED =
        "ScriptErrorHandler requires a CDI observer for ScriptError, but none was found."
        + " Register a bean with a @Observes ScriptError method. See also the javadoc of <o:scriptErrorHandler>.";

    private enum PropertyKeys {
        // Cannot be uppercased. They have to exactly match the attribute names.
        ignoreSelector, maxRecentErrors, errorExpiry;
    }

    // Variables ------------------------------------------------------------------------------------------------------

    private static boolean servletRegistered;

    private final State state = new State(getStateHelper());

    // Actions --------------------------------------------------------------------------------------------------------

    /**
     * Move this component to head, so that error handlers are registered as early as possible.
     * @throws IllegalArgumentException When the {@link ScriptErrorServlet} is not registered because no CDI observer
     * for {@link ScriptError} was found.
     */
    @Override
    public void processEvent(ComponentSystemEvent event) {
        if (event instanceof PostAddToViewEvent) {
            var context = event.getFacesContext();
            validateServletRegistered(context);
            context.getViewRoot().addComponentResource(context, this, "head");
        }
    }

    /**
     * Render the initialization script.
     */
    @Override
    public void encodeChildren(FacesContext context) throws IOException {
        var endpointURL = getRequestContextPath(context) + ScriptErrorServlet.URI;
        var ignoreSelector = getIgnoreSelector();
        var maxRecentErrors = getMaxRecentErrors();
        var errorExpiry = getErrorExpiry();

        var script = format(SCRIPT_INIT,
            endpointURL,
            ignoreSelector != null ? "'" + escapeJS(ignoreSelector, true) + "'" : "null",
            maxRecentErrors,
            errorExpiry);

        context.getResponseWriter().write(script);
    }

    // Attribute getters/setters --------------------------------------------------------------------------------------

    /**
     * Returns the CSS selector for elements whose presence suppresses error reporting.
     * When an element matching this selector exists in the document, errors will not be sent to the server.
     * For example, <code>body.errorPage</code> to suppress reporting on error pages having <code>&lt;h:body styleClass="errorPage"&gt;</code>.
     * @return The CSS selector, or <code>null</code>.
     */
    public String getIgnoreSelector() {
        return state.get(PropertyKeys.ignoreSelector);
    }

    /**
     * Sets the CSS selector for elements whose presence suppresses error reporting.
     * When an element matching this selector exists in the document, errors will not be sent to the server.
     * @param ignoreSelector The CSS selector.
     */
    public void setIgnoreSelector(String ignoreSelector) {
        state.put(PropertyKeys.ignoreSelector, ignoreSelector);
    }

    /**
     * Returns the maximum number of recent errors to keep for deduplication. Defaults to {@value #DEFAULT_MAX_RECENT_ERRORS}.
     * @return The maximum number of recent errors.
     */
    public int getMaxRecentErrors() {
        return state.get(PropertyKeys.maxRecentErrors, DEFAULT_MAX_RECENT_ERRORS);
    }

    /**
     * Sets the maximum number of recent errors to keep for deduplication.
     * Defaults to {@value #DEFAULT_MAX_RECENT_ERRORS}.
     * @param maxRecentErrors The maximum number of recent errors.
     */
    public void setMaxRecentErrors(int maxRecentErrors) {
        state.put(PropertyKeys.maxRecentErrors, maxRecentErrors);
    }

    /**
     * Returns the time in milliseconds after which a recent error entry expires. Defaults to {@value #DEFAULT_ERROR_EXPIRY} (1 minute).
     * @return The error expiry time in milliseconds.
     */
    public int getErrorExpiry() {
        return state.get(PropertyKeys.errorExpiry, DEFAULT_ERROR_EXPIRY);
    }

    /**
     * Sets the time in milliseconds after which a recent error entry expires for deduplication.
     * Defaults to {@value #DEFAULT_ERROR_EXPIRY} (1 minute).
     * @param errorExpiry The error expiry time in milliseconds.
     */
    public void setErrorExpiry(int errorExpiry) {
        state.put(PropertyKeys.errorExpiry, errorExpiry);
    }

    // Helpers --------------------------------------------------------------------------------------------------------

    private static void validateServletRegistered(FacesContext context) {
        if (!servletRegistered) {
            if (!getServletContext(context).getServletRegistrations().containsKey(ScriptErrorServlet.class.getName())) {
                throw new IllegalArgumentException(ERROR_SERVLET_NOT_REGISTERED);
            }

            servletRegistered = true;
        }
    }

    /**
     * Register the script error handler servlet if it has not already been registered and there is at least one CDI
     * observer for {@link ScriptError}.
     * @param servletContext The involved servlet context.
     */
    public static void registerServletIfNecessary(ServletContext servletContext) {
        if (servletContext.getServletRegistrations().containsKey(ScriptErrorServlet.class.getName())) {
            return;
        }

        if (Beans.getManager().resolveObserverMethods(new ScriptError(null, null, null, null, null, null, null, null, null, null)).isEmpty()) {
            return;
        }

        var registration = servletContext.addServlet(ScriptErrorServlet.class.getName(), ScriptErrorServlet.class);

        if (registration != null) {
            registration.addMapping(ScriptErrorServlet.URI);
        }
    }

}

File: src/main/ts/src/ScriptErrorHandler.ts
///
/// Copyright OmniFaces
///
/// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
/// the License. You may obtain a copy of the License at
///
///     https://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
/// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
/// specific language governing permissions and limitations under the License.
///

/**
 * Catch uncaught JavaScript errors and unhandled promise rejections and send them to the server via
 * <code>navigator.sendBeacon()</code>. This script is used by <code>&lt;o:scriptErrorHandler&gt;</code>.
 *
 * @author Bauke Scholtz
 * @see org.omnifaces.component.script.ScriptErrorHandler
 * @since 5.2
 */
export namespace ScriptErrorHandler {

    // "Constant" fields ----------------------------------------------------------------------------------------------

    const MAX_BEACON_LENGTH = 60000; // navigator.sendBeacon() silently refuses a larger payload.
    const TRUNCATION_MARKER = "...";
    const TRIMMABLE_PARAMS = ["errorStack", "errorMessage"]; // In the order they are sacrificed to stay within the maximum.

    // Private static fields ------------------------------------------------------------------------------------------

    const config = {} as { endpointURL: string; ignoreSelector: string | null; maxRecentErrors: number; errorExpiry: number };
    const recentErrors: { key: string; time: number }[] = [];

    // Public static functions ----------------------------------------------------------------------------------------

    /**
     * Initialize the script error handler. This registers <code>window.onerror</code> and
     * <code>unhandledrejection</code> listeners that send error details to the server.
     * @param endpointURL The full URL of the script error handler servlet endpoint.
     * @param ignoreSelector Optional CSS selector; if a matching element exists in the document, the error will not be reported.
     * @param maxRecentErrors Maximum number of recent errors to keep for deduplication.
     * @param errorExpiry Time in milliseconds after which a recent error entry expires.
     */
    export function init(endpointURL: string, ignoreSelector: string | null, maxRecentErrors: number, errorExpiry: number) {
        config.endpointURL = endpointURL;
        config.ignoreSelector = ignoreSelector;
        config.maxRecentErrors = maxRecentErrors;
        config.errorExpiry = errorExpiry;

        window.onerror = (message, source, line, column, error) => {
            send({
                pageURL: window.location.href,
                errorMessage: toErrorMessage(message, error),
                errorName: error?.name ?? null,
                errorStack: error?.stack ?? null,
                sourceURL: source ?? null,
                lineNumber: line != null ? String(line) : null,
                columnNumber: column != null ? String(column) : null
            });

            return false; // Allow the browser's default error handler to continue, e.g. logging to console.
        };

        window.addEventListener("unhandledrejection", (event: PromiseRejectionEvent) => {
            const reason = event.reason;

            send({
                pageURL: window.location.href,
                errorMessage: toReasonMessage(reason),
                errorName: reason instanceof Error ? reason.name : "UnhandledRejection",
                errorStack: reason instanceof Error ? reason.stack ?? null : null
            });
        });
    }

    // Private static functions ---------------------------------------------------------------------------------------

    /**
     * Returns the error message of the given <code>window.onerror</code> arguments. The message is a string when the
     * handler is assigned as property, but a single <code>ErrorEvent</code> carrying the message when it is invoked as
     * event listener, which is how another error handler forwarding to this one would call it. The handler is a public
     * global which anything may call with anything, so every remaining shape falls back to the event type and then to
     * a plain stringification, which cannot throw.
     * @param message The message argument of the onerror handler.
     * @param error The error argument of the onerror handler, if any.
     */
    function toErrorMessage(message: Event | string, error?: Error): string {
        if (typeof message === "string") {
            return message;
        }

        if (error) {
            return error.message;
        }

        if (message instanceof ErrorEvent && message.message) {
            return message.message;
        }

        if (message instanceof Event) {
            return `${message.constructor.name}: ${message.type}`;
        }

        return String(message);
    }

    /**
     * Returns the error message of the given unhandled rejection reason. A rejection commonly carries a plain object
     * holding the interesting detail, which is therefore serialized rather than stringified to "[object Object]".
     * @param reason The reason of the unhandled rejection.
     */
    function toReasonMessage(reason: any): string {
        if (reason instanceof Error) {
            return reason.message;
        }

        if (typeof reason !== "object" || reason === null) {
            return String(reason); // A string reason must stay unquoted.
        }

        try {
            return JSON.stringify(reason) ?? String(reason);
        }
        catch (e) {
            return String(reason); // A circular structure or a BigInt cannot be serialized.
        }
    }

    /**
     * Returns the given error detail parameters as beacon payload, skipping the ones without a value.
     * @param params The error detail parameters.
     */
    function toBeaconData(params: Record<string, string | null>): URLSearchParams {
        const data = new URLSearchParams();

        for (const [name, value] of Object.entries(params)) {
            if (value != null) {
                data.append(name, value);
            }
        }

        return data;
    }

    /**
     * Trims the given error detail parameters so that the encoded payload stays within {@link MAX_BEACON_LENGTH}.
     * Cutting as many characters as the payload is over is always enough, as URL encoding only ever expands.
     * @param params The error detail parameters.
     */
    function trimToMaxLength(params: Record<string, string | null>) {
        for (const name of TRIMMABLE_PARAMS) {
            const excess = toBeaconData(params).toString().length - MAX_BEACON_LENGTH;

            if (excess <= 0) {
                return;
            }

            const value = params[name];

            if (value != null) {
                params[name] = value.substring(0, Math.max(0, value.length - excess - TRUNCATION_MARKER.length)) + TRUNCATION_MARKER;
            }
        }
    }

    /**
     * Send error details to the server via <code>navigator.sendBeacon()</code> with deduplication.
     * @param params The error detail parameters.
     */
    function send(params: Record<string, string | null>) {
        try {
            if (config.ignoreSelector && document.querySelector(config.ignoreSelector)) {
                return;
            }

            trimToMaxLength(params);

            const key = `${params.errorMessage}${params.sourceURL}${params.lineNumber}`;
            const now = Date.now();

            while (recentErrors.length > 0 && now - recentErrors[0].time > config.errorExpiry) {
                recentErrors.shift();
            }

            if (recentErrors.some(entry => entry.key === key)) {
                return;
            }

            if (recentErrors.length >= config.maxRecentErrors) {
                recentErrors.shift();
            }

            recentErrors.push({ key, time: now });
            navigator.sendBeacon(config.endpointURL, toBeaconData(params));
        }
        catch (e) {
            // Fail silently. We don't want to cause additional errors while reporting errors.
        }
    }

}
